### PR TITLE
fix bg_fill abort when server completed

### LINKED ISSUE #1917: Assertion with bg_fill
```
(gdb) p s->state_machine->background_fill
$1 = BACKGROUND_FILL_STARTED


#0  0x00007ffff2be1c37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
        resultvar = 0
        pid = 29943
        selftid = 29948
#1  0x00007ffff2be5028 in __GI_abort () at abort.c:89
        save_stage = 2
        act = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {0, 140737252012688, 
              140737351947559, 5, 0, 0, 140737265737008, 0, 140737252012688, 169, 140737351976149, 17594303951872, 
              17594303953408, 17594303953408, 0, 17594303951872}}, sa_flags = -221632071, sa_restorer = 0x2}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00007ffff4bd5899 in ink_abort (message_format=0x7ffff4c1b080 "%s:%d: failed assertion `%s`") at ink_error.cc:99
        ap = {{gp_offset = 32, fp_offset = 48, overflow_arg_area = 0x7ffff1e9b2a0, reg_save_area = 0x7ffff1e9b1e0}}
#3  0x00007ffff4bcff93 in _ink_assert (expression=0xacc340 "0", file=0xacbe60 "HttpTransact.cc", line=8565) at ink_assert.cc:37
No locals.
#4  0x0000000000744b99 in HttpTransact::update_size_and_time_stats (s=0x7fffeda3c5b8, total_time=18260252794, 
    user_agent_write_time=18186995949, origin_server_read_time=176699, user_agent_request_header_size=89, 
    user_agent_request_body_size=0, user_agent_response_header_size=313, user_agent_response_body_size=0, 
    origin_server_request_header_size=188, origin_server_request_body_size=0, origin_server_response_header_size=169, 
    origin_server_response_body_size=0, pushed_response_header_size=0, pushed_response_body_size=0, milestones=...)
    at HttpTransact.cc:8565
        user_agent_request_size = 89
        origin_server_request_size = 188
        user_agent_response_size = 313
        user_agent_bytes = 402
        origin_server_response_size = 169
        origin_server_bytes = 357
#5  0x00000000006cee3d in HttpSM::update_stats (this=0x7fffeda3c540) at HttpSM.cc:6925
        __FUNCTION__ = "update_stats"
        total_time = 18260252794
        request_process_time = 18233617523
        ua_write_time = 18186995949
        os_read_time = 176699
#6  0x00000000006cd7c8 in HttpSM::kill_this (this=0x7fffeda3c540) at HttpSM.cc:6798
        __FUNCTION__ = "kill_this"
#7  0x00000000006a6c98 in HttpSM::main_handler (this=0x7fffeda3c540, event=2301, data=0x7fffeda3d7d0) at HttpSM.cc:2595
---Type <return> to continue, or q <return> to quit---
        jump_point = NULL
        __FUNCTION__ = "main_handler"
        vc_entry = 0x0
#8  0x0000000000520a9e in Continuation::handleEvent (this=0x7fffeda3c540, event=2301, data=0x7fffeda3d7d0)
    at /root/ats-self/iocore/eventsystem/I_Continuation.h:153
No locals.
#9  0x0000000000763aff in HttpTunnel::main_handler (this=0x7fffeda3d7d0, event=103, data=0x60b00001cce0) at HttpTunnel.cc:1662
        p = 0x0
        c = 0x7fffeda3d8f8
        sm_callback = true
#10 0x0000000000520a9e in Continuation::handleEvent (this=0x7fffeda3d7d0, event=103, data=0x60b00001cce0)
    at /root/ats-self/iocore/eventsystem/I_Continuation.h:153
No locals.
#11 0x0000000000946d33 in CacheVC::calluser (this=0x60b00001ca80, event=103) at P_CacheInternal.h:618
No locals.
#12 0x000000000095fb4c in CacheVC::openWriteMain (this=0x60b00001ca80) at CacheWrite.cc:1312
        called_user = 1
        avail = 200
        ntodo = 200
        total_avail = 200
        towrite = 200
        not_writing = false
#13 0x0000000000520a9e in Continuation::handleEvent (this=0x60b00001ca80, event=1, data=0x608e00004f60)
    at /root/ats-self/iocore/eventsystem/I_Continuation.h:153
No locals.
#14 0x0000000000a20edb in EThread::process_event (this=0x7ffff1e9f800, e=0x608e00004f60, calling_code=1) at UnixEThread.cc:129
        c_temp = 0x60b00001ca80
        lock = {m = {m_ptr = 0x608600025e90}, lock_acquired = true}
        __FUNCTION__ = "process_event"
#15 0x0000000000a213d2 in EThread::execute (this=0x7ffff1e9f800) at UnixEThread.cc:183
        done_one = false
        e = 0x608e00004f60
        NegativeQueue = {<DLL<Event, Event::Link_link>> = {head = 0x608e00006d60}, tail = 0x608e00006b20}
        next_time = 1494922503453636712
        __FUNCTION__ = "execute"
---Type <return> to continue, or q <return> to quit---
#16 0x0000000000a1f519 in spawn_thread_internal (a=0x600800016d90) at Thread.cc:91
        p = 0x600800016d90
#17 0x00007ffff4e63b98 in ?? () from /usr/lib/x86_64-linux-gnu/libasan.so.0
No symbol table info available.
#18 0x00007ffff399f184 in start_thread (arg=0x7ffff1e9c700) at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7ffff1e9c700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140737252017920, 5370358978826929932, 1, 8, 10613713, 140737252017920, 
                -5370337106741754100, -5370333397915465972}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {
              prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
```

### PR DESCRIPTION
**Pull request by scw00**
fixed #1917 

We have already fixed this assertion after #1933 merged , but there are still some problems left.

The bg_fill has broken when Server session released. So we need to check whether the server producer is finished. We set the bg_fill state to BACKGROUND_FILL_COMPLETED if producer completed. Otherwise, the producer is running , the bg_fill can be run as nomal.

### CODE REVIEW

**Comment by scw00** [comment]
@oknet :D

File: proxy/http/HttpSM.cc [hunk 1]
```diff
@@ -3161,15 +3172,25 @@ HttpSM::tunnel_handler_ua(int event, HttpTunnelConsumer *c)
...
+      selfc = p->self_consumer;
+      if (selfc && selfc->producer->vc_type == HT_HTTP_SERVER) {
+        p = selfc->producer;
+      }
+
       DebugSM("http", "[%" PRId64 "] Initiating background fill", sm_id);
-      background_fill = BACKGROUND_FILL_STARTED;
+      // check whether to finish the reading.
+      background_fill = p->read_success ? BACKGROUND_FILL_COMPLETED : BACKGROUND_FILL_STARTED;
       HTTP_INCREMENT_DYN_STAT(http_background_fill_current_count_stat);
```
Comment (oknet) [inline proxy/http/HttpSM.cc]: You forget to decrement the http_background_fill_current_count_stat if BACKGROUND_FILL_COMPLETED

File: proxy/http/HttpSM.cc [hunk 2]
```diff
@@ -3161,15 +3172,25 @@ HttpSM::tunnel_handler_ua(int event, HttpTunnelConsumer *c)
     set_ua_abort(HttpTransact::ABORTED, event);
 
     if (is_bg_fill_necessary(c)) {
+      p     = c->producer;
+      selfc = p->self_consumer;
+      if (selfc && selfc->producer->vc_type == HT_HTTP_SERVER) {
+        p = selfc->producer;
+      }
+
```
Comment (oknet) [inline proxy/http/HttpSM.cc]: Consider rewrite to ```p = c->producer->self_consumer ? c->producer->self_consumer->producer : c->producer;```

**Comment by scw00** [comment]
updated!!

**Comment by zwoop** [comment]
@oknet @scw00 Is this a candidate for v7.1.0 ?

**Comment by scw00** [comment]
IMHO, it's not a serious bug and it's hard to happen.  So 8.0.0 may be better!

**Comment by oknet** [comment]
[approve ci]

**Comment by oknet** [comment]
[approve ci]

**Comment by zwoop** [comment]
[approve ci]

**Comment by oknet** [comment]
@zwoop Is this 2nd reviewer required ?

**Comment by zwoop** [comment]
@oknet No, one reviewer is enough. You should merge it, since no one is objecting.

**Pull request merged**